keras_nlp/tokenizers/word_piece_tokenizer.py

```diff
@@ -0,0 +1,277 @@
+
+# Matches whitespace and control characters.
+WHITESPACE_REGEX = r"|".join(
+    [
+        r"\s",
+        # Invisible control characters
+        r"\p{Cc}",
+        r"\p{Cf}",
+    ]
+)
+
+# Matches punctuation compatible with the original bert implementation.
+PUNCTUATION_REGEX = r"|".join(
+    [
+        # Treat all non-letter/number ASCII as punctuation.
+        # Characters such as "^", "$", and "`" are not in the Unicode
+        # Punctuation class but we treat them as punctuation anyways.
+        r"[!-/]",
+        r"[:-@]",
+        r"[\[-`]",
+        r"[{-~]",
+        # Unicode punctuation class.
+        r"[\p{P}]",
+        # More unicode ranges.
+        r"[\x{4E00}-\x{9FFF}]",
+        r"[\x{3400}-\x{4DBF}]",
+        r"[\x{20000}-\x{2A6DF}]",
+        r"[\x{2A700}-\x{2B73F}]",
+        r"[\x{2B740}-\x{2B81F}]",
+        r"[\x{2B820}-\x{2CEAF}]",
+        r"[\x{F900}-\x{FAFF}]",
+        r"[\x{2F800}-\x{2FA1F}]",
+    ]
+)
+
+# Matches both whitespace and punctuation.
+WHITESPACE_AND_PUNCTUATION_REGEX = r"|".join(
+    [
+        WHITESPACE_REGEX,
+        PUNCTUATION_REGEX,
+    ]
+)
+
+
+class WordPieceTokenizer(tokenizer.Tokenizer):
+    """A word piece tokenizer layer.
+
+    This layer provides an efficient, in graph, implementation of the WordPiece
+    algorithm used by BERT and other models.
+
+    To make this layer more useful out of the box, the layer will pre-tokenize
+    the input, which will optionally lower-case, strip accents, and split the
+    input on whitespace and punctuation. Each of these pre-tokenization steps is
+    not reversible. The `detokenize` method will join words with a space, and
+    will not invert `tokenize` exactly.
+
+    If a more custom pre-tokenization step is desired, the layer can be
+    configured to apply only the strict WordPiece algorithm by passing
+    `lower_case=False`, `strip_accents=False` and `split_pattern=None`. In
+    this case, inputs should be pre-split string tensors or ragged tensors.
+
+    By default, the layer will output a `tf.RaggedTensor` where the last
+    dimension of the output is ragged after whitespace splitting and sub-word
+    tokenizing. If `sequence_length` is set, the layer will output a dense
+    `tf.Tensor` where all inputs have been padded or truncated to
+    `sequence_length`. The output dtype can be controlled via the `dtype`
+    argument, which should be either an integer or string type.
+
+    Args:
+        vocabulary: A list of strings or a string string filename path. If
+            passing a list, each element of the list should be a single word
+            piece token string. If passing a filename, the file should be a
+            plain text file containing a single word piece token per line.
+        sequence_length: If set, the output will be converted to a dense
+            tensor and padded/trimmed so all outputs are of sequence_length.
+        lower_case: If true, the input text will be first lowered before
```

Comment:
Should this be `lowercase`? "lowercase" is more common and is what we use when describing this behavior in `TextVectorization`.

Reply:
Yeah, I think that makes sense. Most of the BERT stuff out there seems to use `do_lower_case` which I think is picked up from the [original BERT implementation](https://github.com/google-research/bert/blob/master/tokenization.py#L164).

But it seems lowercase is the more correct usage https://english.stackexchange.com/questions/59409/lowercase-lower-case-or-lower-case